dttools/src/priority_queue.h

```diff
@@ -0,0 +1,235 @@
+For example, a priority is created with elements with (priority, data) tuples
+(1, "a") (2, "b") (3, "c") (4, "d") (5, "e")
+The order of elements are determined by @ref priority_queue_push.
+The first time to push, the contents are -
+[(1, "a")]
+The second time to push, the priority of the root is less than the new element, the swim operation is applied to the new element -
+[(2, "b"), (1, "a")]
+The third time -
+[(3, "c"), (1, "a"), (2, "b")]
+The fourth time -
+[(4, "d"), (3, "c"), (2, "b"), (1, "a")]
+The fifth time -
+[(5, "e"), (4, "d"), (2, "b"), (1, "a"), (3, "c")]
+As seen, the iteration order of elements is not the same as the priority order.
+
+An example to create a priority queue and manipulate its elements:
+<pre>
+struct priority_queue *pq;
+pq = priority_queue_create(10);
+
+int priority = 5;
+void *data = someDataPointer;
+
+priority_queue_push(pq, data, priority);
+data = priority_queue_pop(pq);
+void *headData = priority_queue_get_head(pq);
+</pre>
+
+To list all of the items in a priority queue, use a simple loop:
+<pre>
+for (int i = 1; i <= priority_queue_size(pq); i++) {
+    void *data = priority_queue_get_element(pq, i);
+    printf("Priority queue contains: %p\n", data);
+}
+</pre>
+
+Or use the PRIORITY_QUEUE_BASE_ITERATE macro:
+
+<pre>
+int idx;
+void *data;
+PRIORITY_QUEUE_BASE_ITERATE (pq, idx, data) {
+    printf("Data idx: %d\n", idx);
+}
+</pre>
+*/
+
+
+/** Create a new priority queue.
+Element with a higher priority is at the top of the heap.
+@param init_capacity The initial number of elements in the queue. If zero, a default value will be used.
+@return A pointer to a new priority queue.
+*/
+struct priority_queue *priority_queue_create(double init_capacity);
+
+/** Count the elements in a priority queue.
+@param pq A pointer to a priority queue.
+@return The number of elements in the queue.
+*/
+int priority_queue_size(struct priority_queue *pq);
+
+/** Push an element into a priority queue.
+The standard push operation. New elements are placed lower than tasks of the same priority
+@param pq A pointer to a priority queue.
+@param data A pointer to store in the queue.
+@param priority The specified priority with the given object.
+@return The idex of data if the push succeeded, -1 on failure.
+*/
+int priority_queue_push(struct priority_queue *pq, void *data, double priority);
+
+/** Push an element into a priority queue.
+New elements are placed upper than tasks of the same priority.
+Used when a task is resubmitted given resource exhaustion, we want it to get to run as soon as possible among tasks of the same priority.
+@param pq A pointer to a priority queue.
+@param data A pointer to store in the queue.
+@param priority The specified priority with the given object.
+@return The idex of data if the push succeeded, -1 on failure.
+*/
+int priority_queue_push_upward(struct priority_queue *pq, void *data, double priority);
```

Comment:
@btovar This version of push was pointed out by @dthain, that when [a task is resubmitted given previous resource exhaustion](https://github.com/cooperative-computing-lab/cctools/blob/a9785efa61520d1acb1a73bbb705068d5bc714d9/taskvine/src/manager/vine_manager.c#L4346C13-L4346C36), we want it to be dispatched as soon as possible in order to get more resources allocated to it. For those big tasks, we don't want to change their assigned priorities, but pushing them upward of those with the same priority is likely to get a bigger chance of scheduling earlier.

Reply:
@JinZhou5042, shouldn't this be the way to push tasks every time then? I don't like that the priority queue abstraction is leaking. If we ever need to change the priority queue because of a future discovered bug, etc., then this will affect the way taskvine schedules tasks, and it will be difficult to figure out why.

Thus, I would remove the other push function and use only this one (after removing the _upward suffix), or increase the priority of elements that need to be pushed again. One option would be to have two priority fields, one as you have it here and another that counts how many times an element has been pushed (e.g. populated with the task try_count from taskvine.)

Reply:
@btovar This makes sense, removing another push function is a better way for abstraction.